modules/view_establishment/predicates.py

```diff
@@ -0,0 +1,344 @@
+        - Returns true if vpair.current is the same as the next view of node j
+        """
+        raise NotImplementedError
+
+    def adopt(self, vpair):
+        """Assign the current view pair as vpair."""
+        raise NotImplementedError
+
+    def establishable(self, phase, mode):
+        """Method description.
+
+        Checks if 4f+1 nodes are willing to move to a view change (phase 0)
+        or to a new view (phase 1).
+        """
+        raise NotImplementedError
+
+    def establish(self):
+        """Update the current view in the view pair to the next view."""
+        raise NotImplementedError
+
+    def next_view(self):
+        """Updates the next view in the view pair to upcoming view."""
+        raise NotImplementedError
+
+    def reset_v_change(self):
+        """The node is no longer in a view change."""
+        raise NotImplementedError
+
+    # Interface functions
+    def need_reset(self):
+        """True if the replication module requires a reset."""
+        return (self.stale_v() and self.resolver.execute(
+                module=Module.REPLICATION_MODULE,
+                func=Function.REPLICA_FLUSH
+                ))
+
+    def reset_all(self):
+        """Reset all modules."""
+        self.views = [self.RST_PAIR for i in range(self.number_of_nodes)]
+        self.view_module.init_module()
+        self.resolver.execute(
+            module=Module.REPLICATION_MODULE,
+            func=Function.REP_REQUEST_RESET
+        )
+        return("Reset")
+
+    def view_change(self):
+        """A view change is required."""
+        self.vChange = True
+
+    def get_view(self, node_j):
+        """Returns the most recent reported view of node_j."""
+        if (node_j == self.id or
+                (self.view_module.get_phs(self.id) == 0 and
+                    self.view_module.witnes_seen())):
+            if self.allow_service():
+                return self.views[self.id].get("current")
+            return self.TEE
+        return self.views[node_j].get("current")
+
+    def allow_service(self):
+        """Method description.
+
+        Returns true if atleast 3f+1 nodes has the same view as
+        current node and current node is not in a view change
+        """
+        return (len(self.same_v_set(self.id)) >
+                3 * self.number_of_byzantine and
+                self.view_module.get_phs(self.id) == 0 and
+                self.views[self.id].get("current") ==
+                self.views[self.id].get("next"))
+
+    def automation(self, type, phase, case):
+        """Perform the action corresponding to the current situation."""
+        # Phase 0, waiting for a next view
+        if (phase == 0):
+            return self.automation_phase_0(type, case)
+
+        # Phase 1, in a view change
+        elif(phase == 1):
+            return self.automation_phase_1(type, case)
+
+        # Not a valid phase
+        else:
+            raise Exception('Not a valid phase: {}'.format(phase))
```

Comment:
Would be nice to raise [ValueError](https://docs.python.org/2/library/exceptions.html#exceptions.ValueError) instead, will make it easier later to have as specific exceptions as possible

Reply:
Done